crates/iceberg/src/expr/visitors/manifest_evaluator.rs

```diff
@@ -0,0 +1,393 @@
+use crate::spec::{Datum, FieldSummary, ManifestFile, PartitionSpecRef, Schema, SchemaRef};
+use fnv::FnvHashSet;
+use std::sync::Arc;
+
+/// Evaluates [`ManifestFile`]s to see if their partition summary matches a provided
+/// [`BoundPredicate`]. Used by [`TableScan`] to filter down the list of [`ManifestFile`]s
+/// in which data might be found that matches the TableScan's filter.
+pub(crate) struct ManifestEvaluator {
+    partition_schema: SchemaRef,
+    partition_filter: BoundPredicate,
+    case_sensitive: bool,
+}
+
+impl ManifestEvaluator {
+    pub(crate) fn new(
+        partition_spec: PartitionSpecRef,
+        table_schema: SchemaRef,
+        partition_filter: BoundPredicate,
+        case_sensitive: bool,
+    ) -> crate::Result<Self> {
+        let partition_type = partition_spec.partition_type(&table_schema)?;
+
+        // this is needed as SchemaBuilder.with_fields expects an iterator over
+        // Arc<NestedField> rather than &Arc<NestedField>
+        let cloned_partition_fields: Vec<_> =
+            partition_type.fields().iter().map(Arc::clone).collect();
+
+        let partition_schema = Schema::builder()
+            .with_fields(cloned_partition_fields)
+            .build()?;
+
+        let partition_schema_ref = Arc::new(partition_schema);
+
+        let mut inclusive_projection = InclusiveProjection::new(partition_spec.clone());
+        let unbound_partition_filter = inclusive_projection.project(&partition_filter)?;
+
+        let partition_filter =
+            unbound_partition_filter.bind(partition_schema_ref.clone(), case_sensitive)?;
+
+        Ok(Self {
+            partition_schema: partition_schema_ref,
+            partition_filter,
+            case_sensitive,
+        })
+    }
+
+    /// Evaluate this `ManifestEvaluator`'s filter predicate against the
+    /// provided [`ManifestFile`]'s partitions. Used by [`TableScan`] to
+    /// see if this `ManifestFile` could possibly contain data that matches
+    /// the scan's filter.
+    pub(crate) fn eval(&self, manifest_file: &ManifestFile) -> crate::Result<bool> {
+        if manifest_file.partitions.is_empty() {
+            return Ok(true);
+        }
+
+        struct ManifestFilterVisitor<'a> {
+            manifest_evaluator: &'a ManifestEvaluator,
+            partitions: &'a Vec<FieldSummary>,
+        }
+
+        impl<'a> ManifestFilterVisitor<'a> {
+            fn new(
+                manifest_evaluator: &'a ManifestEvaluator,
+                partitions: &'a Vec<FieldSummary>,
+            ) -> Self {
+                ManifestFilterVisitor {
+                    manifest_evaluator,
+                    partitions,
+                }
+            }
+        }
+
+        // Remove this annotation once all todos have been removed
+        #[allow(unused_variables)]
+        impl BoundPredicateVisitor for ManifestFilterVisitor<'_> {
+            type T = bool;
+
+            fn always_true(&mut self) -> crate::Result<Self::T> {
```

Comment:
As our previous discussion, we should not concret type rather `Self::T` to make it more readable.

Reply:
Done.